### PR TITLE
[NTOBJSHEX] Sorting improvements

### PR DESCRIPTION
**Pull request by Cleverking2003**
## Purpose

- Add ability to sort NT objects by symlinks.
- Add ability to sort registry entries by type/value.

JIRA issues:
[CORE-18761](https://jira.reactos.org/browse/CORE-18761)
[CORE-18762](https://jira.reactos.org/browse/CORE-18762)

image

### CODE REVIEW

**Comment by KRosUser2** [comment]
Note (out of review of this specific one) : Very similar PR could be raised for https://jira.reactos.org/browse/CORE-18762

File: dll/shellext/ntobjshex/regfolder.cpp [hunk 1]
```diff
@@ -396,11 +396,32 @@ HRESULT CRegistryFolder::CompareIDs(LPARAM lParam, const RegPidlEntry * first, c
...
+
+        HRESULT hr1 = FormatContentsForDisplay(first, m_hRoot, m_NtPath, &firstContent);
+        if (FAILED_UNEXPECTEDLY(hr1))
+            return E_INVALIDARG;
+
+        HRESULT hr2 = FormatContentsForDisplay(second, m_hRoot, m_NtPath, &secondContent);
+        if (FAILED_UNEXPECTEDLY(hr2))
+            return E_INVALIDARG;
+
+        return MAKE_COMPARE_HRESULT(lstrcmpW(firstContent, secondContent));
```
Comment (learn-more) [inline dll/shellext/ntobjshex/regfolder.cpp:423]: Leaking `firstContent`, `secondContent`

File: dll/shellext/ntobjshex/regfolder.cpp [hunk 2]
```diff
@@ -396,11 +396,32 @@ HRESULT CRegistryFolder::CompareIDs(LPARAM lParam, const RegPidlEntry * first, c
...
+
+        return CompareName(lParam, first, second);
 
     case REGISTRY_COLUMN_VALUE:
-        // Can't sort by link target yet
-        return E_INVALIDARG;
+    {
+        PCWSTR firstContent, secondContent;
+
+        HRESULT hr1 = FormatContentsForDisplay(first, m_hRoot, m_NtPath, &firstContent);
```
Comment (learn-more) [inline dll/shellext/ntobjshex/regfolder.cpp:415]: Just `hr` would be fine here?

File: dll/shellext/ntobjshex/regfolder.cpp [hunk 3]
```diff
@@ -396,11 +396,32 @@ HRESULT CRegistryFolder::CompareIDs(LPARAM lParam, const RegPidlEntry * first, c
...
+        if (FAILED_UNEXPECTEDLY(hr1))
+            return E_INVALIDARG;
+
+        HRESULT hr2 = FormatContentsForDisplay(second, m_hRoot, m_NtPath, &secondContent);
```
Comment (learn-more) [inline dll/shellext/ntobjshex/regfolder.cpp:419]: Why not re-use `hr` from above?

File: dll/shellext/ntobjshex/regfolder.cpp [hunk 4]
```diff
@@ -396,11 +396,32 @@ HRESULT CRegistryFolder::CompareIDs(LPARAM lParam, const RegPidlEntry * first, c
...
+        if (FAILED_UNEXPECTEDLY(hr2))
+            return E_INVALIDARG;
```
Comment (learn-more) [inline dll/shellext/ntobjshex/regfolder.cpp:421]: Leaking `firstContent`

File: dll/shellext/ntobjshex/regfolder.cpp [hunk 5]
```diff
@@ -396,11 +396,32 @@ HRESULT CRegistryFolder::CompareIDs(LPARAM lParam, const RegPidlEntry * first, c
         return CompareName(lParam, first, second);
 
     case REGISTRY_COLUMN_TYPE:
-        return MAKE_COMPARE_HRESULT(second->contentType - first->contentType);
+        if (first->entryType != second->entryType)
+            return MAKE_COMPARE_HRESULT(second->entryType - first->entryType);
+
+        if (first->entryType == REG_ENTRY_KEY)
```
Comment (learn-more) [inline dll/shellext/ntobjshex/regfolder.cpp:402]: Other parts of the code also check for `info->contentsLength`

**Review by learn-more — APPROVED** [review]

**Review by JoachimHenze — APPROVED** [review]

File: dll/shellext/ntobjshex/ntobjfolder.cpp [hunk 6]
```diff
@@ -448,8 +448,27 @@ HRESULT CNtObjectFolder::CompareIDs(LPARAM lParam, const NtPidlEntry * first, co
...
+            return E_INVALIDARG;
+
+        RtlInitEmptyUnicodeString(&secondLink, wbLink2, sizeof(wbLink2));
+
+        if (FAILED_UNEXPECTEDLY(GetNTObjectSymbolicLinkTarget(m_NtPath, second->entryName, &secondLink)))
+            return E_INVALIDARG;
+
+        return MAKE_COMPARE_HRESULT(RtlCompareUnicodeString(&firstLink, &secondLink, TRUE));
+    }
     }
```
[the remark this answers is not in the document]
Comment (binarymaster) [reply dll/shellext/ntobjshex/ntobjfolder.cpp]: Ok I have fixed it.

**Review by binarymaster — APPROVED** [review]

**Pull request merged**